File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/filesystem/ROFileSystemProvider.java
package ru.kmorozov.onedrive.filesystem;

import ru.kmorozov.onedrive.CommandLineOpts;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Date;
import java.util.concurrent.TimeUnit;
import java.util.zip.CRC32;
import java.util.zip.CheckedInputStream;

class ROFileSystemProvider implements FileSystemProvider {

    public void delete(File file) throws IOException {
        // Do nothing
    }

    public File createFolder(File file, String name) throws IOException {

        return new File(file, name) {
            @Override
            public boolean isDirectory() {
                return true;
            }
        };
    }

    public File createFile(File file, String name) {
        return new File(file, name);
    }

    public void replaceFile(File original, File replacement) throws IOException {
        // Do nothing
    }

    public void setAttributes(File downloadFile, Date created, Date lastModified) throws IOException {
        // Do nothing
    }

    public boolean verifyCrc(File file, long crc) throws IOException {
        return true;
    }

    public FileSystemProvider.FileMatch verifyMatch(File file, long crc, long fileSize, Date created, Date lastModified) throws IOException {

        // Round to nearest second
        created = new Date((created.getTime() / 1000L) * 1000L);
        lastModified = new Date((lastModified.getTime() / 1000L) * 1000L);

        BasicFileAttributes attr = Files.readAttributes(file.toPath(), BasicFileAttributes.class);

        // Timestamp rounded to the nearest second
        Date localCreatedDate = new Date(attr.creationTime().to(TimeUnit.SECONDS) * 1000L);
        Date localModifiedDate = new Date(attr.lastModifiedTime().to(TimeUnit.SECONDS) * 1000L);

        boolean sizeMatches = fileSize == file.length();
        boolean createdMatches = created.equals(localCreatedDate);
        boolean modifiedMatches = lastModified.equals(localModifiedDate);

        if (!CommandLineOpts.getCommandLineOpts().useHash() && sizeMatches && createdMatches && modifiedMatches) {
            // Close enough!
            return FileSystemProvider.FileMatch.YES;
        }

        long localCrc = this.getChecksum(file);
        boolean crcMatches = crc == localCrc;

        // If the crc matches but the timestamps do not we won't upload the content again
        if (crcMatches && !(modifiedMatches && createdMatches)) {
            return FileSystemProvider.FileMatch.CRC;
        }
        else if (crcMatches) {
            return FileSystemProvider.FileMatch.YES;
        }
        else {
            return FileSystemProvider.FileMatch.NO;
        }
    }

    public FileSystemProvider.FileMatch verifyMatch(File file, Date created, Date lastModified) throws IOException {

        // Round to nearest second
        created = new Date((created.getTime() / 1000L) * 1000L);
        lastModified = new Date((lastModified.getTime() / 1000L) * 1000L);

        BasicFileAttributes attr = Files.readAttributes(file.toPath(), BasicFileAttributes.class);

        // Timestamp rounded to the nearest second
        Date localCreatedDate = new Date(attr.creationTime().to(TimeUnit.SECONDS) * 1000L);
        Date localModifiedDate = new Date(attr.lastModifiedTime().to(TimeUnit.SECONDS) * 1000L);

        boolean createdMatches = created.equals(localCreatedDate);
        boolean modifiedMatches = lastModified.equals(localModifiedDate);

        if (createdMatches && modifiedMatches) {
            return FileSystemProvider.FileMatch.YES;
        }
        else {
            return FileSystemProvider.FileMatch.NO;
        }
    }

    public long getChecksum(File file) throws IOException {
        // Compute CRC32 checksum
        try (final CheckedInputStream cis = new CheckedInputStream(new FileInputStream(file), new CRC32())) {
            byte[] buf = new byte[1024];

            while (0 <= cis.read(buf)) {
            }

            return cis.getChecksum().getValue();
        }
    }
}

File: ru.kmorozov.api.onedrive/src/main/java/ru/kmorozov/onedrive/filesystem/ROFileSystemProvider.kt
package ru.kmorozov.onedrive.filesystem

import ru.kmorozov.onedrive.CommandLineOpts
import ru.kmorozov.onedrive.filesystem.FileSystemProvider.FileMatch

import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.nio.file.Files
import java.nio.file.attribute.BasicFileAttributes
import java.util.Date
import java.util.concurrent.TimeUnit
import java.util.zip.CRC32
import java.util.zip.CheckedInputStream

internal open class ROFileSystemProvider : FileSystemProvider {

    @Throws(IOException::class)
    override fun delete(file: File) {
        // Do nothing
    }

    @Throws(IOException::class)
    override fun createFolder(file: File, name: String): File {

        return object : File(file, name) {
            override fun isDirectory(): Boolean {
                return true
            }
        }
    }

    override fun createFile(file: File, name: String): File {
        return File(file, name)
    }

    @Throws(IOException::class)
    override fun replaceFile(original: File, replacement: File) {
        // Do nothing
    }

    @Throws(IOException::class)
    override fun setAttributes(downloadFile: File, created: Date, lastModified: Date) {
        // Do nothing
    }

    @Throws(IOException::class)
    override fun verifyCrc(file: File, crc: Long): Boolean {
        return true
    }

    @Throws(IOException::class)
    override fun verifyMatch(file: File, crc: Long, fileSize: Long, created: Date, lastModified: Date): FileMatch {
        var created = created
        var lastModified = lastModified

        // Round to nearest second
        created = Date(created.time / 1000L * 1000L)
        lastModified = Date(lastModified.time / 1000L * 1000L)

        val attr = Files.readAttributes(file.toPath(), BasicFileAttributes::class.java)

        // Timestamp rounded to the nearest second
        val localCreatedDate = Date(attr.creationTime().to(TimeUnit.SECONDS) * 1000L)
        val localModifiedDate = Date(attr.lastModifiedTime().to(TimeUnit.SECONDS) * 1000L)

        val sizeMatches = fileSize == file.length()
        val createdMatches = created == localCreatedDate
        val modifiedMatches = lastModified == localModifiedDate

        if (!CommandLineOpts.commandLineOpts.useHash() && sizeMatches && createdMatches && modifiedMatches) {
            // Close enough!
            return FileMatch.YES
        }

        val localCrc = getChecksum(file)
        val crcMatches = crc == localCrc

        // If the crc matches but the timestamps do not we won't upload the content again
        return if (crcMatches && !(modifiedMatches && createdMatches)) {
            FileMatch.CRC
        } else if (crcMatches) {
            FileMatch.YES
        } else {
            FileMatch.NO
        }
    }

    @Throws(IOException::class)
    override fun verifyMatch(file: File, created: Date, lastModified: Date): FileMatch {
        var created = created
        var lastModified = lastModified

        // Round to nearest second
        created = Date(created.time / 1000L * 1000L)
        lastModified = Date(lastModified.time / 1000L * 1000L)

        val attr = Files.readAttributes(file.toPath(), BasicFileAttributes::class.java)

        // Timestamp rounded to the nearest second
        val localCreatedDate = Date(attr.creationTime().to(TimeUnit.SECONDS) * 1000L)
        val localModifiedDate = Date(attr.lastModifiedTime().to(TimeUnit.SECONDS) * 1000L)

        val createdMatches = created == localCreatedDate
        val modifiedMatches = lastModified == localModifiedDate

        return if (createdMatches && modifiedMatches) {
            FileMatch.YES
        } else {
            FileMatch.NO
        }
    }

    @Throws(IOException::class)
    override fun getChecksum(file: File): Long {
        // Compute CRC32 checksum
        CheckedInputStream(FileInputStream(file), CRC32()).use { cis ->
            val buf = ByteArray(1024)

            while (0 <= cis.read(buf)) {
            }

            return cis.checksum.value
        }
    }
}
